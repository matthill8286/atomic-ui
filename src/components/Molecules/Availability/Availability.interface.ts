import * as React from 'react'
import { ProductAvailabilityState } from '@/types/availabilityState'

export interface StyledAvailabilityTypoProps {
  hasIcon: boolean
}

export type TranslatedText = React.ReactNode | string

export type AvailabilitySize = 'small' | 'large'

export interface AvailabilityProps {
  currency?: string
  freeLabel?: string
  loading?: boolean
  infoIconOnClick?: (event: React.MouseEvent<HTMLSpanElement, MouseEvent>) => void
  price?: number | string | null
  scale?: AvailabilitySize
  state?: ProductAvailabilityState
  subtext?: TranslatedText
  text?: TranslatedText
  CustomIcon?: React.MemoExoticComponent<(props: React.SVGProps<SVGSVGElement>) => JSX.Element>
}

export interface AvailabilityStatusProps {
  state?: ProductAvailabilityState
}
