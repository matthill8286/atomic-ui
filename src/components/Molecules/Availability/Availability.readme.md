## Availability molecule

A component to display the availability
