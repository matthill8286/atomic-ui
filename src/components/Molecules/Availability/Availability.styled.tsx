import { Icon } from '@/components/Atoms/Icon'
import { Typo } from '@/components/Atoms/Typography/Typo'
import { styled } from '@/styles/styled'
import { ProductAvailabilityState } from '@/types/availabilityState'
import {
  AvailabilityProps,
  AvailabilityStatusProps,
  StyledAvailabilityTypoProps,
} from './Availability.interface'

export const StyledAvailabilityWrapper = styled.div`
  display: flex;
`
export const StyledAvailabilityStatus = styled.span<AvailabilityStatusProps>`
  display: inline-block;
  height: ${({ theme }) => theme.spacing.base.xs};
  width: ${({ theme }) => theme.spacing.base.xs};
  margin-right: ${({ theme }) => theme.spacing.base.xs};
  border-radius: 50%;

  ${({ theme, state }) => {
    const stateBackgroundMap = new Map<ProductAvailabilityState, string>([
      [ProductAvailabilityState.LOWERTHRESHOLD, `background: ${theme.color.success};`],
      [ProductAvailabilityState.UPPERTHRESHOLD, `background: ${theme.color.success};`],
      [ProductAvailabilityState.AVAILABLE, `background: ${theme.color.success};`],
      [ProductAvailabilityState.PARTIAL_AVAILABLE, `background: ${theme.color.alert};`],
      [ProductAvailabilityState.TIME_CLUSTER, `background: ${theme.color.alert};`],
      [ProductAvailabilityState.RELEASE, `background: ${theme.color.alert};`],
      [ProductAvailabilityState.PICKUP_TIMERANGE, `background: ${theme.color.alert};`],
      [ProductAvailabilityState.NOT_AVAILABLE, `background: ${theme.color.grey3};`],
      [ProductAvailabilityState.NOONLINESTOCK, `background: ${theme.color.grey3};`],
      [ProductAvailabilityState.NOT_POSSIBLE_IN_CITY_CENTER, `background: ${theme.color.grey3};`],
      [ProductAvailabilityState.NOT_POSSIBLE_IN_GENERAL, `background: ${theme.color.grey3};`],
    ])
    return state && stateBackgroundMap.get(state)
  }}
`

export const StyledAvailabilityHeadingWrapper = styled.div`
  flex: 1;
  flex-basis: auto;
`

export const StyledAvailabilityStatusWrapper = styled.span<AvailabilityProps>`
  line-height: ${({ theme, scale }) => scale === 'large' && theme.font.lineHeight.xs};
`

export const StyledAvailabilityPrice = styled.div`
  flex: 0 0 auto;
  color: ${({ theme }) => theme.color.grey5};
  font-family: ${({ theme }) => theme.font.family.default};
  font-weight: 600;
  font-size: ${({ theme }) => theme.font.size.sm};
`

export const StyledAvailabilitySubText = styled.div<AvailabilityProps>`
  background: transparent;
  border: 0;
  margin: 0;
  padding: 0;
  color: ${({ theme }) => theme.color.grey5};
  font-family: ${({ theme }) => theme.font.family.default};
  font-size: ${({ theme, scale }) => (scale == 'large' ? theme.font.size.sm : theme.font.size.xxs)};
`

export const StyledAvailabilityIcon = styled(Icon)`
  cursor: pointer;
  display: inline;
  margin-left: ${({ theme }) => theme.spacing.base.xs};
  position: absolute;
  top: -2px;
  right: ${({ theme }) => `-${theme.spacing.base.md}`};
`

export const StyledAvailabilityTypo = styled(Typo)<StyledAvailabilityTypoProps>`
  position: relative;
  ${({ hasIcon }) => {
    return hasIcon ? `display: inline-flex;` : ''
  }}
`

export const StyledCustomIcon = styled(Icon)`
  vertical-align: middle;
  margin: ${({ theme }) => `${theme.spacing.base.xxs} ${theme.spacing.base.xxs}`} 0 0;
`
