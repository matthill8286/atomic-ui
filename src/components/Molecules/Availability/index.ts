export * from './Availability'
export * from './Availability.interface'
export * from './extractProductAvailabilities'
