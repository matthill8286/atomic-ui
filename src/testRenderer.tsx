import { mount, ReactWrapper } from 'enzyme'
import React from 'react'
import { MemoryRouter as Router } from 'react-router-dom'
import renderer from 'react-test-renderer'
import { ThemeProvider } from 'styled-components'
import { saiyanTheme } from '@/styles/sc-vars-saiyan'
import { Theme } from '@/types'

export function renderWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>,
  theme = saiyanTheme
) {
  return renderer.create(<ThemeProvider theme={theme}>{tree}</ThemeProvider>).toJSON()
}

export function renderWithThemeAndRouter(tree: React.ReactElement, theme: Theme = saiyanTheme) {
  return renderer
    .create(
      <ThemeProvider theme={theme}>
        <Router>{tree}</Router>
      </ThemeProvider>
    )
    .toJSON()
}

export function renderWithRouter(tree: React.ReactElement) {
  return renderer.create(<Router>{tree}</Router>).toJSON()
}

export function mountWithTheme<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(<ThemeProvider theme={saiyanTheme}>{tree}</ThemeProvider>)
}

export function mountWithThemeAndRouter<C extends React.Component, P = C['props'], S = C['state']>(
  tree: React.ReactElement<P>
): ReactWrapper<P, S, C> {
  return mount(
    <ThemeProvider theme={saiyanTheme}>
      <Router>{tree}</Router>
    </ThemeProvider>
  )
}
