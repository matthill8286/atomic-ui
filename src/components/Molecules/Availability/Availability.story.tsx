import { select, text } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'
import { ProductAvailabilityState } from '@/types/availabilityState'
import { Availability } from './Availability'
import { AvailabilityProps, AvailabilitySize } from './Availability.interface'
import Readme from './Availability.readme.md'
import { IconOnlineCourse } from '@matthill8286/atomic-icon-library'

const states: ProductAvailabilityState[] = [
  ProductAvailabilityState.TIME_CLUSTER,
  ProductAvailabilityState.AVAILABLE,
  ProductAvailabilityState.NOT_AVAILABLE,
  ProductAvailabilityState.LOWERTHRESHOLD,
  ProductAvailabilityState.UPPERTHRESHOLD,
  ProductAvailabilityState.NOONLINESTOCK,
]
const prices = [100.99, 'Free', 0]
const scaleOptions = ['large', 'small']

storiesOf('Design System/Molecules/Availability', module)
  .add(
    'Default',
    () => {
      const knobs = (): AvailabilityProps => {
        return {
          scale: select('Type scale', scaleOptions, scaleOptions[0]) as AvailabilitySize,
          state: select('State', states, states[0]),
          text: text('Content', 'Ready to Pickup'),
        }
      }
      return <Availability {...knobs()} />
    },
    { info: Readme }
  )
  .add(
    'With Price',
    () => {
      const knobs = (): AvailabilityProps => {
        return {
          scale: select('Type scale', scaleOptions, scaleOptions[0]) as AvailabilitySize,
          state: select('State', states, states[1]),
          text: text('Content', 'Delivery to your address in 1-3 working days'),
          price: select('Price', prices, prices[0]),
          subtext: text('Content', 'Calculated with Standard Delivery'),
          freeLabel: text('Free Label', 'Gratis'),
        }
      }
      return <Availability {...knobs()} />
    },
    { info: Readme }
  )
  .add(
    'With HTML as subtext',
    () => {
      const knobs = (): AvailabilityProps => {
        return {
          scale: select('Type scale', scaleOptions, scaleOptions[0]) as AvailabilitySize,
          state: select('State', states, states[2]),
          text: text('Content', 'Delivery not available'),
          price: select('Price', prices, prices[0]),
          subtext: <a href="http://mediamarkt.de">MediaMarkt München-Euroindustriepark</a>,
        }
      }
      return <Availability {...knobs()} />
    },
    { info: Readme }
  )
  .add(
    'With availability icon',
    () => {
      const knobs = (): AvailabilityProps => {
        return {
          scale: select('Type scale', scaleOptions, scaleOptions[0]) as AvailabilitySize,
          state: select('State', states, states[0]),
          text: text('Content', 'Ready to Pickup'),
        }
      }
      return <Availability {...knobs()} infoIconOnClick={() => {}} />
    },
    { info: Readme }
  )
  .add(
    'With Icon',
    () => {
      const knobs = (): AvailabilityProps => {
        return {
          scale: select('Type scale', scaleOptions, scaleOptions[0]) as AvailabilitySize,
          state: select('State', states, states[0]),
          text: text('Content', 'Ready to Pickup'),
          subtext: <a href="http://mediamarkt.de">MediaMarkt München-Euroindustriepark</a>,
        }
      }
      return <Availability {...knobs()} CustomIcon={IconOnlineCourse} />
    },
    { info: Readme }
  )
