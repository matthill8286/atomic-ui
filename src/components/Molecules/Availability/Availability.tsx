import React, { ReactNode } from 'react'
import { SkeletonInlineItem } from '@/components/Atoms/Skeleton'
import { convertPrice } from '@/utils/convertPrice'
import { AvailabilityProps, AvailabilitySize } from './Availability.interface'
import {
  StyledAvailabilityHeadingWrapper,
  StyledAvailabilityIcon,
  StyledAvailabilityPrice,
  StyledAvailabilityStatus,
  StyledAvailabilityStatusWrapper,
  StyledAvailabilitySubText,
  StyledAvailabilityTypo,
  StyledAvailabilityWrapper,
  StyledCustomIcon,
} from './Availability.styled'
import { OtherInfo } from '@matthill8286/atomic-icon-library'
import { ProductAvailabilityState } from '@/types'

const textOrNode = (content: ReactNode | string): ReactNode =>
  typeof content === 'string' && /<\/?[a-z][\s\S]*>/.test(content) ? (
    <span dangerouslySetInnerHTML={{ __html: content }} />
  ) : (
    <>{content}</>
  )

const renderSubtext = (
  scale: AvailabilitySize,
  loading: boolean,
  subtext: ReactNode
): JSX.Element | null =>
  !subtext ? null : (
    <StyledAvailabilitySubText scale={scale}>
      {loading ? (
        <SkeletonInlineItem fontSize={scale === 'small' ? 'xxs' : 'sm'} length={20} />
      ) : (
        textOrNode(subtext)
      )}
    </StyledAvailabilitySubText>
  )

export const Availability: React.FC<AvailabilityProps> = ({
  currency,
  freeLabel,
  infoIconOnClick,
  loading = false,
  price,
  scale = 'small',
  state,
  subtext,
  text,
  CustomIcon,
  ...props
}): JSX.Element => {
  return (
    <StyledAvailabilityWrapper {...props}>
      <StyledAvailabilityStatusWrapper scale={scale}>
        {CustomIcon ? (
          <StyledCustomIcon height="sm" width="sm">
            <CustomIcon />
          </StyledCustomIcon>
        ) : (
          <StyledAvailabilityStatus
            state={loading ? ProductAvailabilityState.NOT_AVAILABLE : state}
          />
        )}
      </StyledAvailabilityStatusWrapper>
      <StyledAvailabilityHeadingWrapper>
        {loading ? (
          <SkeletonInlineItem fontSize={scale === 'small' ? 'xxs' : 'sm'} length={12} />
        ) : (
          <StyledAvailabilityTypo
            fontSize={scale === 'small' ? 'xxs' : 'sm'}
            weight="semibold"
            color="grey5"
            hasIcon={!!infoIconOnClick}>
            {textOrNode(text)}
            {infoIconOnClick && (
              <StyledAvailabilityIcon
                as="span"
                color="black"
                height={20}
                onClick={infoIconOnClick}
                width={20}>
                <OtherInfo />
              </StyledAvailabilityIcon>
            )}
          </StyledAvailabilityTypo>
        )}

        {renderSubtext(scale, loading, subtext)}
      </StyledAvailabilityHeadingWrapper>
      <StyledAvailabilityPrice>{convertPrice(price, currency, freeLabel)}</StyledAvailabilityPrice>
    </StyledAvailabilityWrapper>
  )
}

Availability.displayName = 'Availability'
