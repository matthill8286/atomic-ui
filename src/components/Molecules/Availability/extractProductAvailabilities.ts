import { AvailabilityInterface, Product2 } from '@/types/product2'
import { AvailabilityProps } from './Availability.interface'
import { ProductAvailabilityState } from '@/types'

const RE_DELIVERY_SPLIT = /<span>(.*)<\/span>(.*)/i

export const splitDeliveryInformation = (
  value: string,
  defaultText: string
): { text: string; subtext: string | undefined } => {
  const match = RE_DELIVERY_SPLIT.exec(value) || []

  return {
    text: match[1] || defaultText,
    subtext: match[2],
  }
}

export const extractProductAvailabilities = (
  product: Product2,
  noMarketLabel = '',
  defaultOnlineLabel = '',
  defaultMarketLabel = ''
): AvailabilityProps[] => {
  const priceAvailability = {
    market: product?.price?.market || {
      deliveryInformation: noMarketLabel,
      type: '',
      outletId: null,
    },
    online: product?.price?.online || { deliveryInformation: '', type: '' },
  }

  if (!priceAvailability) {
    return []
  }

  const result: AvailabilityProps[] = []

  result.push(availabilityPropsFromProductAvailabily(priceAvailability.online, defaultOnlineLabel))
  result.push(availabilityPropsFromProductAvailabily(priceAvailability.market, defaultMarketLabel))

  return result
}

export const availabilityPropsFromProductAvailabily = (
  availability: AvailabilityInterface,
  defaultText: string
): AvailabilityProps => {
  const { text, subtext } = splitDeliveryInformation(
    availability.deliveryInformation || '',
    defaultText
  )

  return {
    text,
    subtext,
    state:
      (availability.type as ProductAvailabilityState) || ProductAvailabilityState.NOT_AVAILABLE,
  }
}
