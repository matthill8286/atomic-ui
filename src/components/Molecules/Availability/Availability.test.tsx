import * as React from 'react'
import { renderWithTheme } from '@/testRenderer'
import { Availability } from './Availability'
import { ProductAvailabilityState } from '@/types'

describe('Availability', () => {
  const statusContent = 'Available'
  // disabled tests due to failing deployment build
  xit('Should render status time cluster', () => {
    const tree = renderWithTheme(
      <Availability state={ProductAvailabilityState.TIME_CLUSTER}>{statusContent}</Availability>
    )
    expect(tree).toMatchSnapshot()
  })
  xit('Should render price', () => {
    const tree = renderWithTheme(<Availability price="100">{statusContent}</Availability>)
    expect(tree).toMatchSnapshot()
  })
  xit('Should render subtext', () => {
    const tree = renderWithTheme(
      <Availability subtext="Calculated with Standard Delivery">{statusContent}</Availability>
    )
    expect(tree).toMatchSnapshot()
  })
  xit('Should render text', () => {
    const tree = renderWithTheme(
      <Availability text="Delivery to your address in 1-3 working days">
        {statusContent}
      </Availability>
    )
    expect(tree).toMatchSnapshot()
  })
})
