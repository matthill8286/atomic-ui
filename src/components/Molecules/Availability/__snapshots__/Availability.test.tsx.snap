// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Availability Should render price 1`] = `
.c3 {
  margin: 0;
  padding: 0;
  color: #272422;
  font-size: 12px;
  line-height: 16px;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  display: inline-block;
  height: 8px;
  width: 8px;
  margin-right: 8px;
  border-radius: 8px;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 {
  color: #272422;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c4 {
  background: transparent;
  border: 0;
  margin: 4px 0 0 0;
  padding: 0;
  color: #272422;
  font-family: SourceSansPro,Arial,sans-serif;
  font-size: 12px;
}

<div
  className="c0"
>
  <span
    className=""
    scale="small"
  >
    <span
      className="c1"
    />
  </span>
  <div
    className="c2"
  >
    <span
      className="c3"
      color="grey5"
      fontSize="xxs"
    />
    <div
      className="c4"
      scale="small"
    />
  </div>
  <div
    className="c5"
  >
    <span>
      € 100.00
    </span>
  </div>
</div>
`;

exports[`Availability Should render status time cluster 1`] = `
.c3 {
  margin: 0;
  padding: 0;
  color: #272422;
  font-size: 12px;
  line-height: 16px;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  display: inline-block;
  height: 8px;
  width: 8px;
  margin-right: 8px;
  border-radius: 8px;
  background: #ffa550;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 {
  color: #272422;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c4 {
  background: transparent;
  border: 0;
  margin: 4px 0 0 0;
  padding: 0;
  color: #272422;
  font-family: SourceSansPro,Arial,sans-serif;
  font-size: 12px;
}

<div
  className="c0"
>
  <span
    className=""
    scale="small"
  >
    <span
      className="c1"
    />
  </span>
  <div
    className="c2"
  >
    <span
      className="c3"
      color="grey5"
      fontSize="xxs"
    />
    <div
      className="c4"
      scale="small"
    />
  </div>
  <div
    className="c5"
  />
</div>
`;

exports[`Availability Should render subtext 1`] = `
.c3 {
  margin: 0;
  padding: 0;
  color: #272422;
  font-size: 12px;
  line-height: 16px;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  display: inline-block;
  height: 8px;
  width: 8px;
  margin-right: 8px;
  border-radius: 8px;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 {
  color: #272422;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c4 {
  background: transparent;
  border: 0;
  margin: 4px 0 0 0;
  padding: 0;
  color: #272422;
  font-family: SourceSansPro,Arial,sans-serif;
  font-size: 12px;
}

<div
  className="c0"
>
  <span
    className=""
    scale="small"
  >
    <span
      className="c1"
    />
  </span>
  <div
    className="c2"
  >
    <span
      className="c3"
      color="grey5"
      fontSize="xxs"
    />
    <div
      className="c4"
      scale="small"
    >
      Calculated with Standard Delivery
    </div>
  </div>
  <div
    className="c5"
  />
</div>
`;

exports[`Availability Should render text 1`] = `
.c3 {
  margin: 0;
  padding: 0;
  color: #272422;
  font-size: 12px;
  line-height: 16px;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 700;
  -webkit-letter-spacing: normal;
  -moz-letter-spacing: normal;
  -ms-letter-spacing: normal;
  letter-spacing: normal;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c1 {
  display: inline-block;
  height: 8px;
  width: 8px;
  margin-right: 8px;
  border-radius: 8px;
}

.c2 {
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c5 {
  color: #272422;
  font-family: SourceSansPro,Arial,sans-serif;
  font-weight: 600;
  font-size: 16px;
}

.c4 {
  background: transparent;
  border: 0;
  margin: 4px 0 0 0;
  padding: 0;
  color: #272422;
  font-family: SourceSansPro,Arial,sans-serif;
  font-size: 12px;
}

<div
  className="c0"
>
  <span
    className=""
    scale="small"
  >
    <span
      className="c1"
    />
  </span>
  <div
    className="c2"
  >
    <span
      className="c3"
      color="grey5"
      fontSize="xxs"
    >
      Delivery to your address in 1-3 working days
    </span>
    <div
      className="c4"
      scale="small"
    />
  </div>
  <div
    className="c5"
  />
</div>
`;
